sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/firestore/FirestoreV1ReadFn.java

```diff
@@ -109,44 +107,36 @@ protected ServerStreamingCallable<RunQueryRequest, RunQueryResponse> getCallable
     protected RunQueryRequest setStartFrom(
         RunQueryRequest element, RunQueryResponse runQueryResponse) {
       StructuredQuery query = element.getStructuredQuery();
-      StructuredQuery.Builder builder;
-      List<Order> orderByList = query.getOrderByList();
-      // if the orderByList is empty that means the default sort of "__name__ ASC" will be used
-      // Before we can set the cursor to the last document name read, we need to explicitly add
-      // the order of "__name__ ASC" because a cursor value must map to an order by
-      if (orderByList.isEmpty()) {
-        builder =
-            query
-                .toBuilder()
-                .addOrderBy(
-                    Order.newBuilder()
-                        .setField(FieldReference.newBuilder().setFieldPath("__name__").build())
-                        .setDirection(Direction.ASCENDING)
-                        .build())
-                .setStartAt(
-                    Cursor.newBuilder()
-                        .setBefore(false)
-                        .addValues(
-                            Value.newBuilder()
-                                .setReferenceValue(runQueryResponse.getDocument().getName())
-                                .build()));
-      } else {
-        Cursor.Builder cursor = Cursor.newBuilder().setBefore(false);
-        Map<String, Value> fieldsMap = runQueryResponse.getDocument().getFieldsMap();
-        for (Order order : orderByList) {
-          String fieldPath = order.getField().getFieldPath();
-          Value value = fieldsMap.get(fieldPath);
-          if (value != null) {
-            cursor.addValues(value);
-          } else if ("__name__".equals(fieldPath)) {
-            cursor.addValues(
-                Value.newBuilder()
-                    .setReferenceValue(runQueryResponse.getDocument().getName())
-                    .build());
+      StructuredQuery.Builder builder = query.toBuilder();
+      builder.addAllOrderBy(QueryUtils.getImplicitOrderBy(query));
+      Cursor.Builder cursor = Cursor.newBuilder().setBefore(false);
+
+      Map<String, Value> valueMap = runQueryResponse.getDocument().getFieldsMap();
+      for (Order order : builder.getOrderByList()) {
+        List<String> segments;
+        try {
+          segments = QueryUtils.resolveOrderByFieldPath(order.getField().getFieldPath());
+        } catch (IllegalArgumentException e) {
+          // Rethrow as something specific to RunQuery
+          throw new IllegalArgumentException(
+              "Could not retry query due to malformed orderBy field", e.getCause());
+        }
+        // __name__ is a special field and doesn't exist in valueMap
```

Comment:
Is this general firestore knowledge? if not, can we add some sort of documentation link here for the future?

Reply:
The [documentation](https://firebase.google.com/docs/firestore/reference/rest/v1/projects.databases.documents#Document) for `Document` says that __.*__ is reserved, so it sort of implies that __name__ can't be in there. I can link that here.